src/main/java/com/conveyal/taui/grids/GridExporter.java

```diff
@@ -0,0 +1,134 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import spark.Response;
+import static com.conveyal.taui.util.SparkUtil.haltWithJson;
+
+import java.io.IOException;
+import java.io.PipedInputStream;
+import java.io.PipedOutputStream;
+import java.net.URL;
+import java.util.Date;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.zip.GZIPOutputStream;
+
+import static java.lang.Boolean.parseBoolean;
+
+/**
+ * Class with methods to write grids to specified formats, upload to S3, and direct clients to download them.
+ *
+ * Created by ansoncfit on 27-Nov-17, adapting code created by matthewc on 21-Oct-16
+ */
+public class GridExporter {
+
+    private static final Logger LOG = LoggerFactory.getLogger(GridExporter.class);
+
+    private static final ExecutorService executorService = Executors.newSingleThreadExecutor();
+
+    /** How long request URLs are good for */
+    public static final int REQUEST_TIMEOUT_MSEC = 15 * 1000;
+
+    private static void haltWithIncorrectFormat (String format) {
+        haltWithJson(400, "Format \"" + format + "\" is invalid. Request format must be \"grid\", \"png\", or \"tiff\".");
+    }
+
+    /**
+     * Checks two request parameters, ensuring that the requested download format is one of the allowable ones,
+     * and returning a parsed boolean of whether a redirect was requested
+     *
+     * @param redirectText whether to redirect the response
+     * @param format requested format
+     * @return cleaned up version of redirectText
+     */
+    public static boolean checkRedirectAndFormat(String redirectText, String format){
+        if (!"grid".equals(format) && !"png".equals(format) && !"tiff".equals(format)) {
+            haltWithIncorrectFormat(format);
+        }
+
+        boolean redirect;
+        if (redirectText == null || "" .equals(redirectText)) redirect = true;
+        else redirect = parseBoolean(redirectText);
+
+        return redirect;
+    }
+
+    /**
+     * Writes a grid to S3 with requested format
+     *
+     * @param grid grid to write
+     * @param s3 s3 instance
+     * @param bucket name of the bucket
+     * @param key
+     * @param format allowable formats include grid (Conveyal flat binary format), png, and tiff.
+     * @throws IOException
+     */
+    public static void writeToS3(Grid grid, AmazonS3 s3, String bucket, String key, String format) throws IOException {
+        PipedInputStream pis = new PipedInputStream();
+        PipedOutputStream pos = new PipedOutputStream(pis);
+
+        ObjectMetadata om = new ObjectMetadata();
+
+        if ("grid".equals(format)) {
+            om.setContentType("application/octet-stream");
+            om.setContentEncoding("gzip");
+        } else if ("png".equals(format)) {
+            om.setContentType("image/png");
+        } else if ("tiff".equals(format)) {
+            om.setContentType("image/tiff");
+        }
+
+        executorService.execute(() -> {
```

Comment:
Does it matter if this is executed in a Thread if we are going to block for `s3.putObject`?